### ISSUE TITLE
Flake: gcloud crash (connection refused) causes cluster to fail to come up or resource leak

### ISSUE THREAD

**Issue by mikedanese** [kubernetes/kubernetes#24476]
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce/15066/
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce/15063/
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce/15062/
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce/15057/
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce/15056/

gcloud fails API request and either the cluster fails to come up or we leak resources.

@fejta have you seen this?

**Comment by fejta** [comment]
@spxtr @ixdy are there cleanup script problems with the jenkins slaves?

Seems odd that we try to delete the jenkins-e2e-master-https firewall?

However these are all errno 111

http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce/15062/console
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce/15063/console
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce/15057/console
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce/15056/console
```sh
07:48:00 Deleted [https://www.googleapis.com/compute/v1/projects/k8s-jkns-e2e-gce/zones/us-central1-f/instanceGroupManagers/jenkins-e2e-minion-group].
07:48:05 Deleted [https://www.googleapis.com/compute/v1/projects/k8s-jkns-e2e-gce/global/instanceTemplates/jenkins-e2e-minion-template].
07:48:10 Updated [https://www.googleapis.com/compute/v1/projects/k8s-jkns-e2e-gce/zones/us-central1-f/instances/jenkins-e2e-master].
07:49:08 Deleted [https://www.googleapis.com/compute/v1/projects/k8s-jkns-e2e-gce/zones/us-central1-f/instances/jenkins-e2e-master].
07:49:33 Deleted [https://www.googleapis.com/compute/v1/projects/k8s-jkns-e2e-gce/global/firewalls/jenkins-e2e-master-https].
07:49:34 ERROR: gcloud crashed (error): [Errno 111] Connection refused
```

The latest one seems problematic:

http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce/15066/console
```sh
08:54:42 Deleted [https://www.googleapis.com/compute/v1/projects/k8s-jkns-e2e-gce/global/instanceTemplates/jenkins-e2e-minion-template].
08:54:44 ERROR: (gcloud.compute.disks.delete) Some requests did not succeed:
08:54:44  - The disk resource 'jenkins-e2e-master-pd' is already being used by 'jenkins-e2e-master'
```

**Comment by spxtr** [comment]
This is almost certainly related to the build slaves I set up. I'll take a look.

**Comment by spxtr** [comment]
I wouldn't be surprised if the problem is that we're starting up the next job before the cluster has finished tearing down. We recently switched to constantly running this job, not just running it every half an hour. On top of that, I set up the build slaves, which means that we no longer take several minutes to untar kubernetes before starting up the cluster.

**Comment by ixdy** [comment]
We need to figure out what is causing the gcloud crashes. We'd theorized that it was due to either the metadata server issues, or due to the Jenkins VM being overloaded, but we've addressed both of those and are still seeing crashses.

**Comment by mikedanese** [comment]
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce/15084/console

**Comment by spxtr** [comment]
```
15:17:24 2016/04/19 15:17:24 e2e.go:194: Running: Ginkgo tests
15:17:24 Setting up for KUBERNETES_PROVIDER="gke".
15:17:24 ... in gke:prepare-e2e()
15:17:24 ... in gke:detect-master()
15:17:24 ... in gke:detect-project()
15:17:25 ... in gke:detect-node-instance-group()
15:17:25 ERROR: gcloud crashed (error): [Errno 111] Connection refused
15:17:25 
15:17:25 If you would like to report this issue, please run the following command:
15:17:25   gcloud feedback
15:17:26 !!! Error in /workspace/kubernetes/cluster/../cluster/gke/util.sh:276
15:17:26   'NODE_INSTANCE_GROUP=$("${GCLOUD}" ${CMD_GROUP:-} container clusters describe     --project="${PROJECT}" --zone="${ZONE}" "${CLUSTER_NAME}"     | grep instanceGroupManagers | grep "${ZONE}" | cut -d '/' -f 11)' exited with status 1
15:17:26 Call stack:
15:17:26   1: /workspace/kubernetes/cluster/../cluster/gke/util.sh:276 detect-node-instance-group(...)
15:17:26   2: /workspace/kubernetes/hack/ginkgo-e2e.sh:73 main(...)
15:17:26 Exiting with status 1
15:17:26 2016/04/19 15:17:26 e2e.go:200: Error running Ginkgo tests: exit status 1
15:17:26 2016/04/19 15:17:26 e2e.go:196: Step 'Ginkgo tests' finished in 1.146462194s
```
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gke/4643/console

**Comment by ixdy** [comment]
Can we figure out which server gcloud is connecting to when it gets connection refused?

**Comment by bprashanth** [comment]
Maybe we can 
```
$ gcloud compute instance-groups list --verbosity=debug 2> gcloud.log
```
?

And not clobber the log when there's a failure?

**Comment by mikedanese** [comment]
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce/15092/console

**Comment by spxtr** [comment]
Okay, I think I just fixed this, and I hate bash even more than I used to. When I set up the slaves, I started @fejta's metadata cache, added it to `/etc/hosts`, and verified that it was working by pinging it (to make sure it was pointing to the VM's IP) and querying it. Unfortunately, when we made this work with containerized e2e, we [assumed](https://github.com/kubernetes/kubernetes/pull/24281/commits/9fa35d0bc6fc31fb301f9bb3b7e1606db816fbe2#diff-7acb21356e9dacdc0333960fe90f0d69R29) that the entry in `/etc/hosts` would have a space between the IP and "metadata.google.internal", but I used a tab for the slaves, so that entry didn't find its way into the container, so I might as well have been not using the cacher.

I switched them to use spaces, so hopefully it's better now.

**Comment by fejta** [comment]
I'm updating the scripts so we can configure slaves with a single command.

**Comment by spxtr** [comment]
Much better.

**Comment by gmarek** [comment]
It's happening again. Kubemarks aren't running because of metadata servers `Errno 111`.

**Comment by wojtek-t** [comment]
Yup - this is causing all kubemarks runs failing, and thus we are not merging anything.

**Comment by wojtek-t** [comment]
@k8s-oncall - FYI

**Comment by piosz** [comment]
Both `kubernetes-kubemark-5-gce` and `kubernetes-kubemark-500-gce` are continuously failing because of it.

**Comment by piosz** [comment]
This is priority -1. All suites are totally broken because of gcloud crash.

**Comment by dchen1107** [comment]
All critical builds are red since last many hours.

**Comment by dchen1107** [comment]
Can we really enable verbose log for gcloud so that we can really understand what is going on here?

**Comment by zmerlynn** [comment]
@dchen1107: You can see a full stacktrace for `kubemark` [here](http://kubekins.dls.corp.google.com/job/kubernetes-kubemark-5-gce/2538/console). In that case, `kubemark` is failing to refresh its token. I haven't been able to see others, though.

**Comment by zmerlynn** [comment]
I logged onto `jenkins-builder-1` and tried a random gcloud command and it was clearly able to use its token, so it's not that. (Or it's not that builder.)

**Comment by zmerlynn** [comment]
Ah! It looks like `jenkins-builder-2`'s metadata-cache server crashed.

**Comment by bprashanth** [comment]
Just delete the cache? 
169.254.169.254 should still resolve, if it doesn't reboot the vm should fix it.

**Comment by bprashanth** [comment]
By "delete the cache" I mean pop the entry:
```
10.240.0.25 metadata.google.internal
```
in /etc/hosts. I assume that's the cache since it's eth0 of jenkins-2.

**Comment by spxtr** [comment]
We recently created a Jenkins on-call rotation, please ping the appropriate person when this sort of thing happens and let them deal with it.

**Comment by bprashanth** [comment]
I'm guessing you restarted it since I actually see a python process listening on :80. Where is it's init script/logs? (the debug doc in https://github.com/kubernetes/kubernetes/pull/24131/files only lists high level run this /hack/ script instructions, and I didn't look through the script to figure out how it restarts etc).

**Comment by ixdy** [comment]
Also, kubemark isn't running on `jenkins-builder-2`. It's running on `jenkins-master`.

**Comment by piosz** [comment]
@spxtr I pinged you and @fejta via mail.

**Comment by ixdy** [comment]
It looks like the `jenkins-master` metadata server cache server crashed, too.

**Comment by piosz** [comment]
@ixdy @fejta any update? How should we fix this in the future?

**Comment by fejta** [comment]
Once https://github.com/kubernetes/kubernetes/pull/24131 goes in you can use that .sh script to restart the metadata server. After that goes in I'll be improving things to make it automatically restart the cache when it crashes/the machine reboots.

**Comment by piosz** [comment]
Thanks for the explanation.

**Comment by mikedanese** [comment]
This is happening again cc @spxtr @ixdy 

http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-kubemark-5-gce/4921/console

**Comment by mikedanese** [comment]
Multiple suites:

http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gke/7617/console

**Comment by fejta** [comment]
`gcloud compute ssh jenkins-builder-4 gcloud compute instances list` works for all nodes on http://kubekins.dls.corp.google.com/computer/

**Comment by mikedanese** [comment]
Thanks!

**Comment by lavalamp** [comment]
@fejta what was the root cause here? did we do something so this won't reoccur?

**Comment by fejta** [comment]
Root cause was one of the instances doesn't auto restart the metadata server, which I'm fixing. This is tangential to the original problem of this issue. Happy to explain in more detail in a new issue, alternatively see https://github.com/kubernetes/test-infra/pull/63 and https://github.com/kubernetes/test-infra/pull/69

**Comment by quinton-hoole** [comment]
I saw this again today.  

```
+ go run ./hack/e2e.go -v --build
2016/06/10 11:28:43 e2e.go:212: Running: build-release
ERROR: gcloud crashed (error): [Errno 111] Connection refused
```

https://00e9e64bac1bd6278563871399fd4cb81957a382258ab9b033-apidata.googleusercontent.com/download/storage/v1_internal/b/kubernetes-jenkins/o/pr-logs%2Fpull%2F27166%2Fkubernetes-pull-build-test-e2e-gce%2F44155%2Fbuild-log.txt?qk=AD5uMEsvKenoYWy_Xqw3WkEAjmEVGzMcxN3Ayb0e10yZcE4UrIfKntrbIgL6dyZC5-0LDMlE-QhNYpFx79AYa5hkCtLfEYRYLT_botEeR2iEeGtrEqJRxk-6zhr5uJXlRE8pnjGadxndpEaG7vxmpQ83chHWRFQCKXKlFGrC4R0WM43tEPSyLZVx_mFTbH4KfXObr_q82El9eNMQkReT1tSUEsqH8N-08BYb_wW4VkLFNYQMCaF8qgu7dxeFoO8xvMaDE0GBL_M_qId_8m2_J0_5ImkR2lFxolgoBjObeiliXqpPVWhiZ8bkFbfwaw1Qo64Mx2Z7U9jvXyZ7poVkbLIn_Th7OTLUnLg1hSWyjIO0YHpkcjVzsHU5diWBQ7cOzZqsgPWOo1LQA75acMB8TKcmTG8_F8qklvWmpNoG82TWyiZMbORAPVZuJA_a7ZSKr8N75C--9FIhiDZnVDg9H6dVPZTgGH9UCoDB23T1Cyy3gPxmt6QLI7wt9sXREzXb7P454eMbMY2mFRSIfskHj1amepkzJQs7hfYefPagLWdNuyJLeqyRcZRgkOXF4mSItrNKfKFvkp-uj4LCsrrHLK5FpKL2QdKbXs0oQXPgkBhmU5XM20k1MH478AMpNjeaT_csErRKuXkjPe-7pIKsGyhWzrcFiS9Zkf2hGkehXIgQnQRIZUmiKya9W4P7x2tMK8EHVIlozQexPBCj1D6WPJWsIAUGHwlT5j1Tnz6PYaPEUMSCYm45lF50aObhyfI4gKLSMAgAgARn5xN-s9EyfVSWjlD52QwmAIumrY4ru2N66oKSt1Wn6htJFZgAWxQrJTeM6ssxh6-XaATvi-DFxfJPcg6dz5txp2zivkzxYIOT2DS3M9TjJ9cSlb0k9s0Ha_QM2_J8yEStSX115QRSpvKLHpjW0K-KLxJp2Os25DQOlpzLzmmP5mU

**Comment by ixdy** [comment]
The PR Jenkins master VM rebooted again, and the metadata cache server didn't start.

@kubernetes/goog-testing we need to figure out why this isn't starting on boot - P0.

**Comment by janetkuo** [comment]
Again https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/27012/kubernetes-pull-build-test-e2e-gce/44161/artifacts/

**Comment by ixdy** [comment]
yeah, the metadata cache keeps dying.

**Comment by fejta** [comment]
I set the master to 0 executors

**Comment by ixdy** [comment]
https://github.com/kubernetes/test-infra/pull/139 to hopefully help figure out why this keeps crashing.
